README
file2air v1.1 - inject 802.11 packets from binary files <jwright@hasborg.com>


ABOUT

file2air uses the LORCON (http://802.11ninja.net) to inject packets into an
802.11 network, using a binary input file for the transmission information.
I've included a few sample packets in the ./packets/ directory for testing
purposes.

file2air will read the binary input file and transmit the contents onto a
wireless network.  It is assumed the input file is a valid 802.11 packet, with
an appropriate 802.11 header. Through command line options, the user can
override the destination, source and BSSID addresses, sequence number and
other fields with command-line arguments, and can specify the number of
packets to send with an arbitrary delay between each packet.


SETUP

In order to use file2air, you must have LORCON installed on your system,
with a supported wireless card and driver configuration.  See the LORCON
website at http://802.11ninja.net for more information.


GETTING PACKETS

You can create arbitrary packets with any hex editor, but having a starting
point is handy.  Using Wireshark, highlight the frame you want to use as
your starting point, and click on the frame contents.  With the approrpriate
frame bytes highlighted, click File -> Export -> Selected Packet Bytes and
save the contents as a file.  Modify this file as needed (if at all) and
then pass the filename to file2air with the -r argument.


EXAMPLES

- DoS a single user with a repeated deauthentication flood:

  ./file2air -i wlan0 -n 65000 -d 00:01:02:03:04:05 -s 00:40:96:01:02:03 \
             -b 00:40:96:01:02:03 -f ./packets/deauth.bin

  Where 00:01:02:03:04:05 is the victim MAC, and 00:40:96:01:02:03 is the AP
  they are associated with (err, were associated with).

- DoS all the users on an AP:

  ./file2air -i wlan0 -n 65000 -d ff:ff:ff:ff:ff:ff -s 00:40:96:01:02:03 \
             -b 00:40:96:01:02:03 -f ./packets/deauth.bin

- Something much more devious:

  /bin/bash -c "while : ; do
     ./file2air -i wlan0 -n 3 -d ff:ff:ff:ff:ff:ff -s 00:40:96:01:02:03 \
                -b 00:40:96:01:02:03 -f ./packets/deauth.bin
     ./file2air -i wlan0 -n 100 -d ff:ff:ff:ff:ff:ff -s 00:40:96:01:02:03 \
                -b 00:40:96:01:02:03 -f ./packets/proberesp.bin
  done"


CONTACT

Questions, comments, suggestions - jwright@hasborg.com


CREDIT WHERE CREDIT IS DUE

My personal thanks for the following people:

  + FX for the original file2wire souce, which much of this work is based on.
  + Dragorn, for answering my endless questions about C


UNINTERESTING NOTE

Much of this code and this README was written while sitting on the sidewalk in
line for "The Wiggles" tickets.

$Id: README,v 1.4 2005/02/16 13:16:49 jwright Exp $


CHANGELOG
Oct  8 2007  1.1    Added LORCON supported drivers to usage(), released 1.1
Feb 23 2007  1.0RC4 Minor change, fixed -t to be 10/pps not 100/pps.
Feb  8 2007  1.0RC3 Added support for incrementing sequence numbers, and
                    the ability to specify WDS addresses.
Aug 14 2006  1.0RC2 Added logic to not change mode or channel unless necessary
Mar  6 2005  1.0RC1 Updated to use lorcon for packet injection.
Feb 16 2005  0.4    Bug fixes, using wlan-ng properly now.  Updated KoreK patch
                    for wlan-ng 0.2.1-pre26 drivers, working on 2.6.10 kernel.
Feb  5 2005  0.3    Modified to use wlan-ng drivers for packet injection.
Sep 20 2003  0.1    Initial release.


crc.c
/*
 * file2air - injects binary files as packets onto 802.11 networks
 *
 * $Id: crc.c,v 1.1 2005/03/16 11:23:41 jwright Exp $
 *
 * Copyright (c) 2005, Joshua Wright <jwright@hasborg.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation. See COPYING for more
 * details.
 *
 * file2air is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * Many thanks for the original file2wire source to:
 * FX <fx@phenoelit.de>
 * Phenoelit (http://www.phenoelit.de)
 * (c) 2k1
 *
 */

#include "crc.h"

unsigned long crc_table[] = {
	0x00000000, 0x77073096, 0xee0e612c, 0x990951ba, 0x076dc419, 0x706af48f,
	0xe963a535, 0x9e6495a3, 0x0edb8832, 0x79dcb8a4, 0xe0d5e91e, 0x97d2d988,
	0x09b64c2b, 0x7eb17cbd, 0xe7b82d07, 0x90bf1d91, 0x1db71064, 0x6ab020f2,
	0xf3b97148, 0x84be41de, 0x1adad47d, 0x6ddde4eb, 0xf4d4b551, 0x83d385c7,
	0x136c9856, 0x646ba8c0, 0xfd62f97a, 0x8a65c9ec, 0x14015c4f, 0x63066cd9,
	0xfa0f3d63, 0x8d080df5, 0x3b6e20c8, 0x4c69105e, 0xd56041e4, 0xa2677172,
	0x3c03e4d1, 0x4b04d447, 0xd20d85fd, 0xa50ab56b, 0x35b5a8fa, 0x42b2986c,
	0xdbbbc9d6, 0xacbcf940, 0x32d86ce3, 0x45df5c75, 0xdcd60dcf, 0xabd13d59,
	0x26d930ac, 0x51de003a, 0xc8d75180, 0xbfd06116, 0x21b4f4b5, 0x56b3c423,
	0xcfba9599, 0xb8bda50f, 0x2802b89e, 0x5f058808, 0xc60cd9b2, 0xb10be924,
	0x2f6f7c87, 0x58684c11, 0xc1611dab, 0xb6662d3d, 0x76dc4190, 0x01db7106,
	0x98d220bc, 0xefd5102a, 0x71b18589, 0x06b6b51f, 0x9fbfe4a5, 0xe8b8d433,
	0x7807c9a2, 0x0f00f934, 0x9609a88e, 0xe10e9818, 0x7f6a0dbb, 0x086d3d2d,
	0x91646c97, 0xe6635c01, 0x6b6b51f4, 0x1c6c6162, 0x856530d8, 0xf262004e,
	0x6c0695ed, 0x1b01a57b, 0x8208f4c1, 0xf50fc457, 0x65b0d9c6, 0x12b7e950,
	0x8bbeb8ea, 0xfcb9887c, 0x62dd1ddf, 0x15da2d49, 0x8cd37cf3, 0xfbd44c65,
	0x4db26158, 0x3ab551ce, 0xa3bc0074, 0xd4bb30e2, 0x4adfa541, 0x3dd895d7,
	0xa4d1c46d, 0xd3d6f4fb, 0x4369e96a, 0x346ed9fc, 0xad678846, 0xda60b8d0,
	0x44042d73, 0x33031de5, 0xaa0a4c5f, 0xdd0d7cc9, 0x5005713c, 0x270241aa,
	0xbe0b1010, 0xc90c2086, 0x5768b525, 0x206f85b3, 0xb966d409, 0xce61e49f,
	0x5edef90e, 0x29d9c998, 0xb0d09822, 0xc7d7a8b4, 0x59b33d17, 0x2eb40d81,
	0xb7bd5c3b, 0xc0ba6cad, 0xedb88320, 0x9abfb3b6, 0x03b6e20c, 0x74b1d29a,
	0xead54739, 0x9dd277af, 0x04db2615, 0x73dc1683, 0xe3630b12, 0x94643b84,
	0x0d6d6a3e, 0x7a6a5aa8, 0xe40ecf0b, 0x9309ff9d, 0x0a00ae27, 0x7d079eb1,
	0xf00f9344, 0x8708a3d2, 0x1e01f268, 0x6906c2fe, 0xf762575d, 0x806567cb,
	0x196c3671, 0x6e6b06e7, 0xfed41b76, 0x89d32be0, 0x10da7a5a, 0x67dd4acc,
	0xf9b9df6f, 0x8ebeeff9, 0x17b7be43, 0x60b08ed5, 0xd6d6a3e8, 0xa1d1937e,
	0x38d8c2c4, 0x4fdff252, 0xd1bb67f1, 0xa6bc5767, 0x3fb506dd, 0x48b2364b,
	0xd80d2bda, 0xaf0a1b4c, 0x36034af6, 0x41047a60, 0xdf60efc3, 0xa867df55,
	0x316e8eef, 0x4669be79, 0xcb61b38c, 0xbc66831a, 0x256fd2a0, 0x5268e236,
	0xcc0c7795, 0xbb0b4703, 0x220216b9, 0x5505262f, 0xc5ba3bbe, 0xb2bd0b28,
	0x2bb45a92, 0x5cb36a04, 0xc2d7ffa7, 0xb5d0cf31, 0x2cd99e8b, 0x5bdeae1d,
	0x9b64c2b0, 0xec63f226, 0x756aa39c, 0x026d930a, 0x9c0906a9, 0xeb0e363f,
	0x72076785, 0x05005713, 0x95bf4a82, 0xe2b87a14, 0x7bb12bae, 0x0cb61b38,
	0x92d28e9b, 0xe5d5be0d, 0x7cdcefb7, 0x0bdbdf21, 0x86d3d2d4, 0xf1d4e242,
	0x68ddb3f8, 0x1fda836e, 0x81be16cd, 0xf6b9265b, 0x6fb077e1, 0x18b74777,
	0x88085ae6, 0xff0f6a70, 0x66063bca, 0x11010b5c, 0x8f659eff, 0xf862ae69,
	0x616bffd3, 0x166ccf45, 0xa00ae278, 0xd70dd2ee, 0x4e048354, 0x3903b3c2,
	0xa7672661, 0xd06016f7, 0x4969474d, 0x3e6e77db, 0xaed16a4a, 0xd9d65adc,
	0x40df0b66, 0x37d83bf0, 0xa9bcae53, 0xdebb9ec5, 0x47b2cf7f, 0x30b5ffe9,
	0xbdbdf21c, 0xcabac28a, 0x53b39330, 0x24b4a3a6, 0xbad03605, 0xcdd70693,
	0x54de5729, 0x23d967bf, 0xb3667a2e, 0xc4614ab8, 0x5d681b02, 0x2a6f2b94,
	0xb40bbe37, 0xc30c8ea1, 0x5a05df1b, 0x2d02ef8d
};

unsigned long crc32(unsigned char *buffer, int buffer_len) {
	int i;
	unsigned long crc=0xffffffff;

	for(i=0; i<buffer_len; i++) {
		crc = crc_table[(crc ^ buffer[i]) & 0xff] ^ (crc >>8);
	}

	return(crc);
}




crc.h
/*
 * file2air - injects binary files as packets onto 802.11 networks
 *
 * $Id: crc.h,v 1.1 2005/03/16 11:23:41 jwright Exp $
 *
 * Copyright (c) 2005, Joshua Wright <jwright@hasborg.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation. See COPYING for more
 * details.
 *
 * file2air is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * Many thanks for the original file2wire source to:
 * FX <fx@phenoelit.de>
 * Phenoelit (http://www.phenoelit.de)
 * (c) 2k1
 *
 */

unsigned long crc32(unsigned char *buffer,int buffer_len);



utils.c

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <crypt.h>
#include <unistd.h>
#include <ctype.h>
#include <netinet/in.h>

/* A better version of hdump, from Lamont Granquist.  Modified slightly
   by Fyodor (fyodor@DHP.com) */
void lamont_hdump(unsigned char *bp, unsigned int length) {

  /* stolen from tcpdump, then kludged extensively */

  static const char asciify[] = "................................ !\"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~.................................................................................................................................";

  const unsigned short *sp;
  const unsigned char *ap;
  unsigned int i, j;
  int nshorts, nshorts2;
  int padding;

  printf("\n\t");
  padding = 0;
  sp = (unsigned short *)bp;
  ap = (unsigned char *)bp;
  nshorts = (unsigned int) length / sizeof(unsigned short);
  nshorts2 = (unsigned int) length / sizeof(unsigned short);
  i = 0;
  j = 0;
  while(1) {
    while (--nshorts >= 0) {
      printf(" %04x", ntohs(*sp));
      sp++;
      if ((++i % 8) == 0)
        break;
    }
    if (nshorts < 0) {
      if ((length & 1) && (((i-1) % 8) != 0)) {
        printf(" %02x  ", *(unsigned char *)sp);
        padding++;
      }
      nshorts = (8 - (nshorts2 - nshorts));
      while(--nshorts >= 0) {
        printf("     ");
      }
      if (!padding) printf("     ");
    }
    printf("  ");

    while (--nshorts2 >= 0) {
      printf("%c%c", asciify[*ap], asciify[*(ap+1)]);
      ap += 2;
      if ((++j % 8) == 0) {
        printf("\n\t");
        break;
      }
    }
    if (nshorts2 < 0) {
      if ((length & 1) && (((j-1) % 8) != 0)) {
        printf("%c", asciify[*ap]);
      }
      break;
    }
  }
  if ((length & 1) && (((i-1) % 8) == 0)) {
    printf(" %02x", *(unsigned char *)sp);
    printf("                                       %c", asciify[*ap]);
  }
  printf("\n");
}


int IsBlank(char *s) {

    int len, i;
    if (s == NULL) { 
        return(1); 
    }

    len = strlen(s);

    if (len == 0) { 
        return(1); 
    }

    for (i=0; i < len; i++) {
       if (s[i] != ' ') {
           return(0);
       }
    }
    return(0);
}


void to_upper (char *s) {
    char	*p;
    char	offset;

    offset = 'A' - 'a';
    for(p=s;*p != '\0';p++) {
        if(islower(*p)) {
            *p += offset;
        }
    }
}


/*
 * converts a string to a mac address...
 * returns 0 on success, -1 on failure...
 * failure indicates poorly formed input...
 */
int string_to_mac (char *string, unsigned char *mac_buf) {
    char			*ptr, *next;
    unsigned long	val;
    int				i;


    to_upper(string);

    ptr = next = string;
    for(i=0;i < 6;i++) {
        if((val = strtoul(next, &ptr, 16)) > 255) {
            return(-1);
        }
        mac_buf[i] = (unsigned char)val;
        if((next == ptr) && (i != 6 - 1)) {
            return(-1);
        }
        next = ptr + 1;
    }

    return(0);
}



common.h
/*
 * file2air - injects binary files as packets onto 802.11 networks
 *
 * $Id: common.h,v 1.1 2005/03/16 11:23:41 jwright Exp $
 *
 * Copyright (c) 2005, Joshua Wright <jwright@hasborg.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation. See COPYING for more
 * details.
 *
 * file2air is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * Many thanks for the original file2wire source to:
 * FX <fx@phenoelit.de>
 * Phenoelit (http://www.phenoelit.de)
 * (c) 2k1
 *
 */

typedef unsigned char u8;
typedef unsigned short u16;
typedef unsigned long u32;


file2air.h
/*
 * file2air - injects binary files as packets onto 802.11 networks
 *
 * $Id: file2air.h,v 1.8 2007/02/08 20:47:52 jwright Exp $
 *
 * Copyright (c) 2005, Joshua Wright <jwright@hasborg.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation. See COPYING for more
 * details.
 *
 * file2air is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * Many thanks for the original file2wire source to:
 * FX <fx@phenoelit.de>
 * Phenoelit (http://www.phenoelit.de)
 * (c) 2k1
 *
 */

#define MAXCHANNEL 11
#define MAXPACKETSIZE 2312
#define MINPACKETSIZE 10
#define A4MINLEN 30
#define PROGNAME "file2air"
#define VER "1.1"

struct optcfg {
    char		device[IFNAMSIZ+1];
    int			verbose;
    char		filename[1024];
    int			mode;
    int			monitor;
    int			channel;
    int			count;
    char		usleep_s[16+1];
    int			usleep;
    int         fragpieces;
};

struct ieee80211 {
    u8    version:2;
    u8    type:2;
    u8    subtype:4;
    u8    to_ds:1;
    u8    from_ds:1;
    u8    more_frag:1;
    u8    retry:1;
    u8    pwrmgmt:1;
    u8    more_data:1;
    u8    wep:1;
    u8    order:1;
    u16	  duration;
    u8    addr1[6];
    u8    addr2[6];
    u8    addr3[6];
    u16   fragment:4;
    u16   sequence:12;
} __attribute__ ((packed));

struct ieee80211a4 {
    u8    version:2;
    u8    type:2;
    u8    subtype:4;
    u8    to_ds:1;
    u8    from_ds:1;
    u8    more_frag:1;
    u8    retry:1;
    u8    pwrmgmt:1;
    u8    more_data:1;
    u8    wep:1;
    u8    order:1;
    u16	  duration;
    u8    addr1[6];
    u8    addr2[6];
    u8    addr3[6];
    u8    addr4[6];
    u16   fragment:4;
    u16   sequence:12;
} __attribute__ ((packed));

/* Prototypes */
void usage(char *n);
int sendpackets(struct tx80211 *in_tx, struct optcfg *cfg, 
                unsigned char *packet, int plen);


file2air.c
/*
* file2air - injects binary files as packets onto 802.11 networks
*
* $Id: file2air.c,v 1.9 2007/02/08 20:47:52 jwright Exp $
*
* Copyright (c) 2005, Joshua Wright <jwright@hasborg.com>
*
* This program is free software; you can redistribute it and/or modify
* it under the terms of the GNU General Public License version 2 as
* published by the Free Software Foundation. See COPYING for more
* details.
*
* file2air is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* Many thanks for the original file2wire source to:
* FX <fx@phenoelit.de>
* Phenoelit (http://www.phenoelit.de)
* (c) 2k1
*
*/

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <ctype.h>
#include <sys/socket.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <getopt.h>

#include <tx80211.h>
#include <tx80211_packet.h>

#include <sys/socket.h>
#include <linux/wireless.h>

#include "common.h"
#include "utils.h"
#include "file2air.h"


void usage(char *message) {

	struct tx80211_cardlist *cardlist = NULL;
	int i;

	cardlist = tx80211_getcardlist();


	if (strlen(message) > 0) {
		printf("%s: %s\n", PROGNAME, message);
	}

	printf("Usage: %s [options]\n", PROGNAME);
	printf("\n"
	"  -i  --interface\tSpecify an interface name\n"
	"  -r  --driver   \tDriver type for injection\n"
	"  -f  --filename \tSpecify a binary file contents for injection\n"
	"\n"
	"  -c  --channel  \tChannel number\n"
	"  -n  --count	\tNumber of packets to send\n"
	"  -w  --delay	\tDelay between packets (uX for usec or X for seconds)\n"
	"  -t  --fast	 \tAlias for -w u100000 (10 packets per second)\n"
	"\n"
	"  -d  --dest	 \tOverride the destination address\n"
	"  -s  --source   \tOverride the source address\n"
	"  -b  --bssid	\tOverride the BSSID address\n"
	"  -a  --wds	  \tOverride the WDS address\n"
	"  -q  --seqnum   \tOverride the sequence number (leading 0x for hex value)\n"
	"  -Q  --seqnuminc\tOverride the sequence number and increment sequentially\n"
	"  -p  --pieces   \tFragment the payload into X pieces.\n"
	"\n"
	"  -h  --help	 \tOutput this help information and exit\n"
	"  -v  --verbose  \tPrint verbose info (more -v's for more verbosity)\n"
	"\n");
	
	if (cardlist == NULL) {
		fprintf(stderr, "Error accessing supported cardlist.\n");
	} else {
		printf("\nSupported drivers are: ");
		for (i = 1; i < cardlist->num_cards; i++) {
			printf("%s ", cardlist->cardnames[i]);
		}
		printf("\n");
	}

}


/* returns -1 on error, number of bytes read on success */
int readfile(unsigned char *buffer, struct optcfg *cfg) {

	int fd;
	struct stat sfilebuf;

	if (stat(cfg->filename, &sfilebuf) !=0) {
		perror("stat");
		return(1);
	} 

	if (cfg->verbose) {
		printf("%s - %ld bytes raw data\n", cfg->filename,
		sfilebuf.st_size);
	}

	if (sfilebuf.st_size > MAXPACKETSIZE) {
		fprintf(stderr, "Packet size too large (%d).  Must be %d or "
		"smaller.\n", (int)sfilebuf.st_size,
		MAXPACKETSIZE);
		return(-1);
	}

	if ((fd = open(cfg->filename, O_RDONLY)) < 0) {
		perror("open()");
		return (1);
	}

	read(fd, buffer, sfilebuf.st_size);
	close(fd);

	return(sfilebuf.st_size);
}

int sendpackets(struct tx80211 *in_tx, struct optcfg *cfg, 
		unsigned char *packet, int plen) {

	int ret, i, payloadlen, fragsize, lastfragsize, wholefrags, offset;
	unsigned char *fragpacket;
	struct tx80211_packet in_packet;
	struct ieee80211 *dot11ptr;

	tx80211_initpacket(&in_packet);

	if (cfg->fragpieces > 1) {
		/* fragment packet for transmission */
		payloadlen = plen - (sizeof(struct ieee80211));

		/* Ensure payload length is greater then the number of 
		fragments */
		if (!(payloadlen >= cfg->fragpieces)) 
		return -2;

		/* Calculate the size of fragments, and the last fragment 
		size */
		fragsize = payloadlen / cfg->fragpieces;
		if (payloadlen % cfg->fragpieces) {
			lastfragsize = (payloadlen - 
			(fragsize * (cfg->fragpieces - 1)));
		} else {
			lastfragsize = 0;
		}

		/* Offset of the packet to transmit as fragments, starting at
		payload */
		offset=sizeof(struct ieee80211);

		/* Number of fragments to send */
		if (lastfragsize > 0) {
			wholefrags = cfg->fragpieces-1;
		} else {
			wholefrags = cfg->fragpieces;
		}

		/* Specify the packet size to transmit */
		in_packet.plen = sizeof(struct ieee80211)+fragsize;

		/* Packet to transmit with specified header contents and 
		   fragmented payload contents. */
		fragpacket=malloc(sizeof(struct ieee80211)+fragsize);
		if (fragpacket == NULL) {
			return -3;
		}
		memcpy(fragpacket, packet, sizeof(struct ieee80211));
		dot11ptr = (struct ieee80211 *)fragpacket;

		/* Set the More Fragments flag in the Frame Control header */
		dot11ptr->more_frag=1;

		/* Send fragments */
		for (i=0; i < wholefrags; i++) {
			dot11ptr->fragment = i;
			memcpy(fragpacket+sizeof(struct ieee80211), 
					packet+offset, fragsize);
			in_packet.packet = fragpacket;

			/* Test for last fragment and no final partiaa
			   fragment */
			if(i == wholefrags-1 && lastfragsize == 0) {
				dot11ptr->more_frag=0;
			}

			ret = tx80211_txpacket(in_tx, &in_packet);
			if (ret < (sizeof (struct ieee80211)+fragsize)) {
				return -4;
			}
			offset+=fragsize;
		}
		free(fragpacket);

		/* Send last fragment */
		if (lastfragsize != 0) {

			/* Packet to transmit with specified header contents 
				and fragmented payload contents. */
			fragpacket=malloc(sizeof(struct ieee80211) +
			lastfragsize);
			if (fragpacket == NULL) {
				return -3;
			}
			memcpy(fragpacket, packet, sizeof(struct ieee80211));

			dot11ptr = (struct ieee80211 *)fragpacket;
			dot11ptr->fragment = i;
			dot11ptr->more_frag = 0;

			memcpy(fragpacket+sizeof(struct ieee80211), 
			packet + offset, lastfragsize);

			in_packet.packet = fragpacket;
			in_packet.plen = sizeof(struct ieee80211)+lastfragsize;

			ret = tx80211_txpacket(in_tx, &in_packet);
			if (ret < (sizeof (struct ieee80211)+lastfragsize)) {
				return -6;
			}

			free(fragpacket);
		}

		ret = plen;

	} else {

		/* Setup the tx80211_packet struct to transmit */
		in_packet.plen = plen;
		in_packet.packet = packet;
		ret = tx80211_txpacket(in_tx, &in_packet);

	}

	return ret;
}

int getseqnum(char *seqnum) {
	int ret=0;
	if (strncmp(seqnum, "0x", 2) == 0) {
		sscanf(seqnum+2, "%x", &ret);
	} else {
		sscanf(seqnum, "%d", &ret);
	}

	return ret;
}

int main(int argc,char *argv[]) {
	char option, opt_dst[18], opt_src[18], opt_bssid[18],
			opt_wds[18], opt_seqnum[8];
	extern char *optarg;
	struct optcfg cfg;
	int opt_srcset=0, opt_dstset=0, opt_bssidset=0,
			opt_wdsset=0, opt_seqnumset=0, opt_seqnumincset=0;
	int ret, plen, seqnum, i;
	unsigned char buffer[MAXPACKETSIZE];
	unsigned char src[6], dst[6], bssid[6], wds[6];
	struct ieee80211 *dot11ptr;
	struct ieee80211a4 *dot11ptra4;
	struct tx80211 in_tx;
	int drivertype = INJ_NODRIVER;

	memset(&cfg, 0, sizeof(cfg));
	/* setup defaults */
	cfg.count=1;

	printf("file2air v%s - inject 802.11 packets from binary files "
	"<jwright@hasborg.com>\n", VER);

	/* Cribbed from the GNU docs on getopt_long */
	while (1) {
		static struct option long_options[] =
		{
			{"interface", required_argument, 0, 'i'},
			{"filename", required_argument, 0, 'f'},
			{"count", required_argument, 0, 'n'},
			{"delay", required_argument, 0, 'w'},
			{"source", required_argument, 0, 's'},
			{"bssid", required_argument, 0, 'b'},
			{"dest", required_argument, 0, 'd'},
			{"wds", required_argument, 0, 'a'},
			{"help", no_argument, 0, 'h'},
			{"verbose", no_argument, 0, 'v'},
			{"driver", required_argument, 0, 'r'},
			{"channel", required_argument, 0, 'c'},
			{"seqnum", required_argument, 0, 'q'},
			{"seqnuminc", required_argument, 0, 'Q'},
			{"fast", no_argument, 0, 't'},
			{"pieces", required_argument, 0, 'p'},
			{0, 0, 0, 0}
		};
		/* getopt_long stores the option index here. */
		int option_index = 0;

		option = getopt_long (argc, argv, 
				"c:i:f:n:w:s:b:d:a:hvr:q:Q:tp:",
				long_options, &option_index);

		/* Detect the end of the options. */
		if (option == -1)
		break;

		switch (option) {
			case 'v':	
			cfg.verbose++;
			break;
			case 'c':
			cfg.channel = atoi(optarg);
			break;
			case 'i':	
			strncpy(cfg.device, optarg, sizeof(cfg.device)-1);
			break;
			case 'f':	
			strncpy(cfg.filename, optarg, sizeof(cfg.filename)-1);
			break;
			case 'n':   
			cfg.count = atoi(optarg);
			break;
			case 'w':   
			strncpy(cfg.usleep_s, optarg, sizeof(cfg.usleep_s)-1);
			break;
			case 's':  
			strncpy(opt_src, optarg, sizeof(opt_src)-1);
			opt_srcset=1;
			break;
			case 'd':   
			strncpy(opt_dst, optarg, sizeof(opt_dst)-1);
			opt_dstset=1;
			break;
			case 'b':   
			strncpy(opt_bssid, optarg, sizeof(opt_bssid)-1);
			opt_bssidset=1;
			break;
			case 'a':
			strncpy(opt_bssid, optarg, sizeof(opt_wds)-1);
			opt_wdsset=1;
			break;
			case 'r':
			drivertype = tx80211_resolvecard(optarg);
			break;
			case 'q':
			strncpy(opt_seqnum, optarg, sizeof(opt_seqnum)-1);
			opt_seqnumset=1;
			break;
			case 'Q':
			strncpy(opt_seqnum, optarg, sizeof(opt_seqnum)-1);
			opt_seqnumset=1;
			opt_seqnumincset=1;
			break;
			case 't':
			strncpy(cfg.usleep_s, "u100000", 
					sizeof(cfg.usleep_s)-1);
			break;
			case 'p':
			cfg.fragpieces = atoi(optarg);
			break;
			default:	
			usage("");
			exit(1);
		}
	}

	if (*cfg.device == 0 || *cfg.filename == 0) {
		usage("Must specify -i and -f");
		return(1);
	}

	if (drivertype == INJ_NODRIVER) {
		usage("Driver name not recognized.\n");
		return(1);
	}

	/* Initialize */
	if (tx80211_init(&in_tx, cfg.device, drivertype) < 0) {
		perror("tx80211_init");
		return 1;
	}

	/* Open the interface to get a socket */
	ret = tx80211_open(&in_tx);
	if (ret < 0) {
		fprintf(stderr, "Unable to open interface %s: %s.\n", 
		cfg.device, in_tx.errstr);
		return 1;
	}

	/* Set packet injection mode */
	ret = tx80211_setfunctionalmode(&in_tx, TX80211_FUNCMODE_INJECT);
	if (ret != 0) {
		fprintf(stderr, "Error setting mode, returned %d.\n", ret);
		return 1;
	}

	/* Switch to the given channel */
	if (cfg.channel != 0) {
		ret = tx80211_setchannel(&in_tx, cfg.channel);
		if (ret < 0) {
			fprintf(stderr, "Error setting channel, returned %d."
					"\n", ret);
			return 1;
		}
	}

	/* Amount of time to sleep between packets.  If we get uXXX, treat that 
	   as a sleep time in usec, otherwise assume it is in seconds  */
	if (*cfg.usleep_s != 0) {
		if (*cfg.usleep_s == 'u') {
			cfg.usleep_s[0] = '0';
			cfg.usleep = atoi(cfg.usleep_s);
		} else {
			cfg.usleep = (atoi(cfg.usleep_s) * 1000000);
		}
	} else {
		cfg.usleep = 0;
	}


	/* Test address parameters to detect malformed MAC's */
	if (opt_srcset) {
		if (string_to_mac(opt_src, src) != 0) {
			usage("Invalid src address.");
			return(1);
		}
	}

	if (opt_dstset) {
		if (string_to_mac(opt_dst, dst) != 0) {
			usage("Invalid dst address.");
			return(1);
		}
	}

	if (opt_bssidset) {
		if (string_to_mac(opt_bssid, bssid) != 0) {
			usage("Invalid bssid address.");
			return(1);
		}
	}

	if (opt_wdsset) {
		if (string_to_mac(opt_wds, wds) != 0) {
			usage("Invalid WDS address.");
			return(1);
		}
	}

	/* Read file to inject, storing contents in buffer, return packet 
	   length */
	plen = readfile(buffer, &cfg);
	if (plen < MINPACKETSIZE) {
		fprintf(stderr, "Error reading input file %s.\n", cfg.filename);
		return(-1);
	}

	dot11ptr = (struct ieee80211 *)&buffer[0];
	dot11ptra4 = (struct ieee80211a4 *)&buffer[0];

	/* If we passed alternative source, destination or bssid addresses at 
	the command line, copy them over the addresses used in the input file. 
	XXX This is broken, since it does not take into considertation the
	status of the From and To DS flags.  For practical purposes, it 
	assumes that we are injecting packets as a wireless station to the AP */
	if (opt_dstset) {
		memcpy(dot11ptr->addr1, dst, sizeof(dot11ptr->addr1));
	}

	if (opt_srcset) {
		memcpy(dot11ptr->addr2, src, sizeof(dot11ptr->addr2));
	}

	if (opt_bssidset) {
		memcpy(dot11ptr->addr3, bssid, sizeof(dot11ptr->addr3));
	}

	if (opt_seqnumset) {
		/* Test driver for capability to specify sequence
		   number */
		if ((tx80211_getcapabilities(&in_tx) & TX80211_CAP_SEQ)
				== 0) {
			fprintf(stderr, "Driver does not support "
					"specifying a "
					"sequence number.\n");
			return -1;
		}

		seqnum = getseqnum(opt_seqnum);

		/* test for invalid sequence number > 4095 */
		if (seqnum > 4095) {
			fprintf(stderr, "Invalid sequence numberi "
					"(max 4095).\n");
			return(-1);
		}

	}

	if (cfg.fragpieces > 0) {
		/* Fragment payload into the specified number of
		   fragments */
		if ((tx80211_getcapabilities(&in_tx) &
				TX80211_CAP_FRAG) == 0) {
			fprintf(stderr, "Driver does not support "
					"fragmentation.\n");
			return -1;
		}
	}

	if (opt_wdsset) {
		if (plen < A4MINLEN) {
			fprintf(stderr, "Frame length (%d) too small for WDS."
					"\n", plen);
			return(-1);
		} else {
			memcpy(dot11ptra4->addr4, wds, 
					sizeof(dot11ptra4->addr4));
			if (opt_seqnumset) dot11ptra4->sequence = seqnum;
		}

	} else { /* Assume 3-address header */
		if (opt_seqnumset) dot11ptr->sequence = seqnum;

	}


	/* Send the data as a frame */
	printf("Transmitting packets ... ");
	fflush(stdout);
	for (i=0; i < cfg.count; i++) {

		if (cfg.verbose) {
			lamont_hdump(buffer, plen);
			printf("Packet length: %d\n", plen);
		}
		ret = sendpackets(&in_tx, &cfg, buffer, plen);
		if (ret < plen) {
			fprintf(stderr, "Error transmitting packet: %s\n",
					in_tx.errstr);
			return -1;
		}
		usleep(cfg.usleep);

		/* Support for spoofing sequential sequence numbers */
		if (opt_seqnumincset) {
			dot11ptr->sequence++;
			if (dot11ptr->sequence > 4095) {
				dot11ptr->sequence=0;
			}
		}
	}

	/* Close the socket */
	tx80211_close(&in_tx);

	printf("Done\n");

	return(0);
}

